File: ts/project/.sdk/tm/java/feature/ProxyFeature.java
package JAVAPACKAGE.feature;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import JAVAPACKAGE.core.Context;
import JAVAPACKAGE.core.SdkClient;
import JAVAPACKAGE.core.Utility;

// Outbound HTTP(S) proxy support. Wraps the active transport and annotates
// each request's fetch definition with the proxy target (`fetchdef.proxy`).
// The default HttpClient transport honours the annotation by routing the
// request through a proxied client (see utility/Fetcher.java); custom
// transports can do the same. The proxy target comes from options (`url`)
// or, when `fromEnv` is set, the standard HTTPS_PROXY / HTTP_PROXY /
// NO_PROXY environment variables. Hosts matching `noProxy` bypass the proxy.
@SuppressWarnings({"unchecked"})
public class ProxyFeature extends BaseFeature {

  private SdkClient client;
  private Map<String, Object> options;
  private List<String> noProxy = new ArrayList<>();

  // Activity tracking (mirrors the ts client._proxy record).
  public int routed = 0;
  public String url = "";

  private static final Pattern HOST_RE =
      Pattern.compile("^[a-z]+://([^/:]+)", Pattern.CASE_INSENSITIVE);

  public ProxyFeature() {
    super("proxy", "0.0.1", true);
  }

  @Override
  public void init(Context ctx, Map<String, Object> options) {
    this.client = ctx.client;
    this.options = options;
    this.active = FeatureOptions.foptBool(options, "active", false);

    if (!this.active) {
      return;
    }

    this.url = FeatureOptions.foptStr(this.options, "url", "");
    List<String> noProxyRaw = FeatureOptions.foptStrList(this.options, "noProxy");

    if (FeatureOptions.foptBool(this.options, "fromEnv", false)) {
      if ("".equals(this.url)) {
        this.url = firstEnv("HTTPS_PROXY", "https_proxy", "HTTP_PROXY", "http_proxy");
      }
      if (noProxyRaw == null) {
        String np = firstEnv("NO_PROXY", "no_proxy");
        if (!"".equals(np)) {
          noProxyRaw = new ArrayList<>(List.of(np.split(",")));
        }
      }
    }

    this.noProxy = new ArrayList<>();
    if (noProxyRaw != null) {
      for (String np : noProxyRaw) {
        np = np.trim();
        if (!"".equals(np)) {
          this.noProxy.add(np);
        }
      }
    }

    final Utility.FetcherFn inner = ctx.utility.fetcher;

    ctx.utility.fetcher = (ctx2, u, fetchdef) ->
        inner.fetch(ctx2, u, route(u, fetchdef));
  }

  private Map<String, Object> route(String u, Map<String, Object> fetchdef) {
    if ("".equals(this.url) || bypass(u)) {
      return fetchdef;
    }

    Map<String, Object> out = new LinkedHashMap<>(fetchdef);
    out.put("proxy", this.url);

    this.routed++;
    return out;
  }

  private boolean bypass(String u) {
    if (this.noProxy.isEmpty()) {
      return false;
    }
    String host = u;
    Matcher m = HOST_RE.matcher(u);
    if (m.find()) {
      host = m.group(1);
    }
    for (String np : this.noProxy) {
      if ("*".equals(np)) {
        return true;
      }
      String suffix = np.startsWith(".") ? np.substring(1) : np;
      if (host.equals(np) || host.endsWith("." + suffix)) {
        return true;
      }
    }
    return false;
  }

  private static String firstEnv(String... names) {
    for (String name : names) {
      String v = System.getenv(name);
      if (v != null && !"".equals(v)) {
        return v;
      }
    }
    return "";
  }
}

File: ts/project/.sdk/tm/scala/feature/ProxyFeature.scala
package SCALAPACKAGE.feature

import java.util.{ArrayList, LinkedHashMap, List => JList, Map => JMap}
import java.util.regex.{Matcher, Pattern}
import SCALAPACKAGE.core.{Context, FetcherFn, SdkClient}

// Outbound HTTP(S) proxy support. Wraps the active transport and annotates
// each request's fetch definition with the proxy target (`fetchdef.proxy`).
// The default HttpClient transport honours the annotation by routing the
// request through a proxied client (see utility/Fetcher.java); custom
// transports can do the same. The proxy target comes from options (`url`)
// or, when `fromEnv` is set, the standard HTTPS_PROXY / HTTP_PROXY /
// NO_PROXY environment variables. Hosts matching `noProxy` bypass the proxy.
class ProxyFeature extends BaseFeature("proxy", "0.0.1", true) {

  private var client: SdkClient = null
  private var options: JMap[String, Object] = null
  private var noProxy: JList[String] = new ArrayList[String]()

  // Activity tracking (mirrors the ts client._proxy record).
  var routed: Int = 0
  var url: String = ""

  private val HOST_RE: Pattern =
    Pattern.compile("^[a-z]+://([^/:]+)", Pattern.CASE_INSENSITIVE)

  override def init(ctx: Context, options: JMap[String, Object]): Unit = {
    this.client = ctx.client
    this.options = options
    this.active = FeatureOptions.foptBool(options, "active", false)

    if (!this.active) {
      return
    }

    this.url = FeatureOptions.foptStr(this.options, "url", "")
    var noProxyRaw = FeatureOptions.foptStrList(this.options, "noProxy")

    if (FeatureOptions.foptBool(this.options, "fromEnv", false)) {
      if ("".equals(this.url)) {
        this.url = firstEnv("HTTPS_PROXY", "https_proxy", "HTTP_PROXY", "http_proxy")
      }
      if (noProxyRaw == null) {
        val np = firstEnv("NO_PROXY", "no_proxy")
        if (!"".equals(np)) {
          noProxyRaw = new ArrayList[String](java.util.List.of(np.split(",")*))
        }
      }
    }

    this.noProxy = new ArrayList[String]()
    if (noProxyRaw != null) {
      val it = noProxyRaw.iterator()
      while (it.hasNext) {
        val np = it.next().trim()
        if (!"".equals(np)) {
          this.noProxy.add(np)
        }
      }
    }

    val inner: FetcherFn = ctx.utility.fetcher

    ctx.utility.fetcher = (ctx2, u, fetchdef) => inner(ctx2, u, route(u, fetchdef))
  }

  private def route(u: String, fetchdef: JMap[String, Object]): JMap[String, Object] = {
    if ("".equals(this.url) || bypass(u)) {
      return fetchdef
    }

    val out = new LinkedHashMap[String, Object](fetchdef)
    out.put("proxy", this.url)

    this.routed += 1
    out
  }

  private def bypass(u: String): Boolean = {
    if (this.noProxy.isEmpty) {
      return false
    }
    var host = u
    val m = HOST_RE.matcher(u)
    if (m.find()) {
      host = m.group(1)
    }
    val it = this.noProxy.iterator()
    while (it.hasNext) {
      val np = it.next()
      if ("*".equals(np)) {
        return true
      }
      val suffix = if (np.startsWith(".")) np.substring(1) else np
      if (host.equals(np) || host.endsWith("." + suffix)) {
        return true
      }
    }
    false
  }

  private def firstEnv(names: String*): String = {
    val it = names.iterator
    while (it.hasNext) {
      val name = it.next()
      val v = System.getenv(name)
      if (v != null && !"".equals(v)) {
        return v
      }
    }
    ""
  }
}
